include/circt/Dialect/Calyx/CalyxControl.td

```diff
@@ -68,6 +68,7 @@ def IfOp : CalyxContainer<"if", [
 
   let assemblyFormat = "$cond (`with` $groupName^)? $thenRegion (`else` $elseRegion^)? attr-dict";
   let verifier = "return ::verify$cppClass(*this);";
+  let hasCanonicalizeMethod = true;
```

Comment:
Why choose `hasCanonicalizeMethod` + explicit canonicalization function over the `hasCanonicalizer` + canonicalization patterns approach?

Reply:
No strong reason. Any reason why you prefer one over the other?

Reply:
I'd imagine `hasCanonicalizeMethod` is only to be preferred where:
1. An explicit order is required in the canonicalization method
2. There is a need to do complex transformations/logic which cannot be captured by the pattern rewrite driver.

Reply:
Is this stated somewhere?

I mostly just read through [this](https://mlir.llvm.org/docs/Canonicalization/#canonicalizing-with-rewritepatterns), and it seemed the `hasCanonicalizeMethod` was the simplest of the two.

Reply:
I think `hasCanonicalizeMethod` makes sense to use whenever the situation is simple enough to fit into one simple pattern. If more flexibility is needed, `hasCanonicalizer` can be more general.